controllers/remote/cluster_cache.go

```diff
@@ -54,319 +48,206 @@ const (
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating client and cache for remote cluster")
+	}
 
-	// EventHandler contains the event handlers to invoke for resource events.
-	EventHandler handler.EventHandler
+	t.clusterAccessors[cluster] = a
 
-	// Predicates is used to filter resource events.
-	Predicates []predicate.Predicate
+	return a, nil
 }
 
-// Watch watches a remote cluster for resource events. If the watch already exists based on cluster, watcher,
-// kind, and eventHandler, then this is a no-op.
-func (m *ClusterCacheTracker) Watch(ctx context.Context, input WatchInput) error {
-	gvk, err := apiutil.GVKForObject(input.Kind, m.scheme)
+// newClusterAccessor creates a new clusterAccessor.
+func (t *ClusterCacheTracker) newClusterAccessor(ctx context.Context, cluster client.ObjectKey) (*clusterAccessor, error) {
+	// Get a rest config for the remote cluster
+	config, err := RESTConfig(ctx, t.client, cluster)
 	if err != nil {
-		return err
-	}
-
-	wi := watchInfo{
-		watcher:               input.Watcher,
-		gvk:                   gvk,
-		eventHandlerSignature: eventHandlerSignature(input.EventHandler),
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
 	}
+	config.Timeout = defaultClientTimeout
 
-	// First, check if the watch already exists
-	var exists bool
-	m.watchesLock.RLock()
-	exists = m.watchExists(input.Cluster, wi)
-	m.watchesLock.RUnlock()
-
-	if exists {
-		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
-		return nil
+	// Create a mapper for it
+	mapper, err := apiutil.NewDynamicRESTMapper(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating dynamic rest mapper for remote cluster")
 	}
 
-	// Doesn't exist - grab the write lock
-	m.watchesLock.Lock()
-	defer m.watchesLock.Unlock()
-
-	// Need to check if another goroutine created the watch while this one was waiting for the lock
-	if m.watchExists(input.Cluster, wi) {
-		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
-		return nil
+	// Create the client for the remote cluster
+	c, err := client.New(config, client.Options{Scheme: t.scheme, Mapper: mapper})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating client for remote cluster")
 	}
 
-	// Need to create the watch
-	watchesForCluster, found := m.watches[input.Cluster]
-	if !found {
-		watchesForCluster = make(map[watchInfo]struct{})
-		m.watches[input.Cluster] = watchesForCluster
+	// Create the cache for the remote cluster
+	cacheOptions := cache.Options{
+		Scheme: t.scheme,
+		Mapper: mapper,
 	}
-
-	cache, err := m.getOrCreateClusterCache(ctx, input.Cluster)
+	remoteCache, err := cache.New(config, cacheOptions)
 	if err != nil {
-		return err
+		return nil, errors.Wrap(err, "error creating cache for remote cluster")
```

Comment:
Let's maybe add more info to the wrap? Like the remote cluster name?

Reply:
Just saw this - will do!